microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceASTValidator.java

```diff
@@ -226,50 +255,76 @@ private void validateAsynchronousAnnotation(MethodDeclaration node, MarkerAnnota
+			Expression jitterExpr = getAnnotationMemberValueExpression(
+					annotation, JITTER_RETRY_ANNOTATION_MEMBER);
+			Expression jitterUnitExpr = getAnnotationMemberValueExpression(
+					annotation, JITTER_DELAY_UNIT_RETRY_ANNOTATION_MEMBER);
 
 			int delayNum = (int) delayExpr.resolveConstantExpressionValue();
-			int maxDurationNum = (int) maxDurationExpr.resolveConstantExpressionValue();
-			int jitterNum = jitterExpr != null ? (int) jitterExpr.resolveConstantExpressionValue() : 0;
+			int maxDurationNum = (int) maxDurationExpr
+					.resolveConstantExpressionValue();
+			int jitterNum = jitterExpr != null
+					? (int) jitterExpr.resolveConstantExpressionValue()
+					: 0;
+
+			Duration delayValue = findDurationUnit(delayUnitExpr, delayNum);
+			Duration maxDurationValue = findDurationUnit(durationUnitExpr,
+					maxDurationNum);
+			Duration jitterValue = findDurationUnit(jitterUnitExpr, jitterNum);
 
-			String delayUnit = delayUnitExpr != null ? delayUnitExpr.toString().split("\\.")[1] : null;
-			String maxDurationUnit = durationUnitExpr != null ? durationUnitExpr.toString().split("\\.")[1] : null;
-			String jitterDelayUnit = jitterDelayUnitExpr != null ? jitterDelayUnitExpr.toString().split("\\.")[1]
-					: null;
+			if (delayValue == null || maxDurationValue == null
+					|| jitterValue == null) {
+				return;
+			}
 
-			Duration delayValue = delayUnitExpr != null ? Duration.of(delayNum, ChronoUnit.valueOf(delayUnit))
-					: Duration.of(delayNum, ChronoUnit.MILLIS);
-			Duration maxDurationValue = durationUnitExpr != null
-					? Duration.of(maxDurationNum, ChronoUnit.valueOf(maxDurationUnit))
-					: Duration.of(maxDurationNum, ChronoUnit.MILLIS);
-			Duration jitterValue = jitterDelayUnitExpr != null
-					? Duration.of(jitterNum, ChronoUnit.valueOf(jitterDelayUnit))
-					: Duration.of(jitterNum, ChronoUnit.MILLIS);
 			Duration maxDelayValue = delayValue.plus(jitterValue);
 
 			if (maxDelayValue.compareTo(maxDurationValue) >= 0) {
-				super.addDiagnostic(RETRY_ERROR_MESSAGE, DIAGNOSTIC_SOURCE, delayExpr, DELAY_EXCEEDS_MAX_DURATION,
-						DiagnosticSeverity.Error);
+				super.addDiagnostic(RETRY_WARNING_MESSAGE, DIAGNOSTIC_SOURCE,
+						delayExpr, DELAY_EXCEEDS_MAX_DURATION,
+						DiagnosticSeverity.Warning);
 			}
 		}
 	}
 
-	private boolean isAllowedReturnTypeForAsynchronousAnnotation(String returnType) {
+	private Duration findDurationUnit(Expression memberUnitExpr,
+			int memberUnitNum) {
+		String memberUnit = null;
+		if (memberUnitExpr != null) {
+			SimpleName memberUnitName = memberUnitExpr instanceof SimpleName
+					? (SimpleName) memberUnitExpr
+					: ((QualifiedName) memberUnitExpr).getName();
+			memberUnit = memberUnitName.getIdentifier();
+		}
+		Duration memberValue = null;
+		try {
+			memberValue = memberUnit != null
+					? Duration.of(memberUnitNum, ChronoUnit.valueOf(memberUnit))
+					: Duration.of(memberUnitNum, ChronoUnit.MILLIS);
+		} catch (UnsupportedTemporalTypeException e) {

```

Comment:
I believe this would be a syntax error on the java side then.

i.e. consider what you entered:
```java
@Retry(delay = 1, maxDuration = 1, delayUnit = ChronoUnit.HOU)
```
Then the client will have the following error:
```
HOU cannot be resolved or is not a fieldJava(33554502)
```

If this is the case, do we still need to handle illegal arguments?

Reply:
If we do handle the illegal arguments, I assume it covers type mismatch, cannot be resolved to variable, etc..